File: hallucinate_app/node/ipfs_datasets.js
import { ipfsDatasetsJs } from 'ipfs_datasets_js';
import { ipfsKitJs } from 'ipfs_kit_js';
import path from 'path';
import fs from 'fs';
import os from 'os';

/**
 * Enhanced IPFS Datasets module
 * Extends the base ipfs_datasets_js functionality with additional methods
 */
class IPFSDatasets {
  constructor(resources = {}, metadata = {}) {
    // Initialize base instance
    this.ipfsDatasetsJs = new ipfsDatasetsJs(resources, metadata);
    
    // Store resources and metadata
    this.resources = resources;
    this.metadata = metadata;
    
    // Get IPFS Kit if available
    this.ipfsKit = resources.ipfsKit || (resources.ipfs_kit ? resources.ipfs_kit : null);
    
    // Get default dataset from metadata
    this.defaultDataset = metadata.dataset || null;
    
    // Get datasets path from metadata or default
    this.datasetsPath = metadata.datasetsPath || path.join(os.homedir(), '.cache', 'huggingface', 'datasets');
    
    // Ensure datasets path exists
    if (!fs.existsSync(this.datasetsPath)) {
      fs.mkdirSync(this.datasetsPath, { recursive: true });
    }
    
    // Dataset registry to track local datasets
    this.datasetRegistry = {};
    this.registryLoaded = false;
    
    console.log(`IPFS Datasets initialized with path=${this.datasetsPath}`);
  }
  
  /**
   * Initialize the datasets module
   */
  async init() {
    try {
      // Initialize base instance
      if (this.ipfsDatasetsJs.init) {
        await this.ipfsDatasetsJs.init();
      }
      
      // Initialize IPFS Kit if available
      if (this.ipfsKit && this.ipfsKit.init) {
        await this.ipfsKit.init();
      }
      
      // Load registry
      await this.loadRegistry();
      
      // Load default dataset if specified
      if (this.defaultDataset) {
        await this.loadDataset(this.defaultDataset);
      }
      
      console.log('IPFS Datasets initialized successfully');
      return true;
    } catch (error) {
      console.error('IPFS Datasets initialization failed:', error);
      return false;
    }
  }
  
  /**
   * Call method on the base ipfs_datasets_js instance
   * @param {string} method - Method name
   * @param {object} args - Method arguments
   */
  async call(method, args) {
    if (!this.ipfsDatasetsJs.call) {
      throw new Error('Base ipfs_datasets_js instance does not have call method');
    }
    
    return await this.ipfsDatasetsJs.call(method, args);
  }
  
  /**
   * Load dataset registry from local storage
   */
  async loadRegistry() {
    const registryPath = path.join(this.datasetsPath, 'dataset_registry.json');
    
    if (fs.existsSync(registryPath)) {
      try {
        const data = fs.readFileSync(registryPath, 'utf8');
        this.datasetRegistry = JSON.parse(data);
        this.registryLoaded = true;
        
        console.log(`Loaded ${Object.keys(this.datasetRegistry).length} datasets from registry`);
        return true;
      } catch (error) {
        console.error('Failed to load dataset registry:', error);
        this.datasetRegistry = {};
        return false;
      }
    } else {
      console.log('No dataset registry found, creating new one');
      this.datasetRegistry = {};
      this.registryLoaded = true;
      return true;
    }
  }
  
  /**
   * Save dataset registry to local storage
   */
  async saveRegistry() {
    const registryPath = path.join(this.datasetsPath, 'dataset_registry.json');
    
    try {
      fs.writeFileSync(registryPath, JSON.stringify(this.datasetRegistry, null, 2), 'utf8');
      console.log(`Saved ${Object.keys(this.datasetRegistry).length} datasets to registry`);
      return true;
    } catch (error) {
      console.error('Failed to save dataset registry:', error);
      return false;
    }
  }
  
  /**
   * List all datasets in the registry
   */
  async listDatasets() {
    if (!this.registryLoaded) {
      await this.loadRegistry();
    }
    
    return this.datasetRegistry;
  }
  
  /**
   * Get information about a specific dataset
   * @param {string} datasetId - Dataset identifier
   */
  async getDatasetInfo(datasetId) {
    if (!this.registryLoaded) {
      await this.loadRegistry();
    }
    
    return this.datasetRegistry[datasetId];
  }
  
  /**
   * Load a dataset using base ipfs_datasets_js implementation
   * @param {string} datasetId - Dataset identifier
   * @param {object} options - Additional options
   */
  async loadDataset(datasetId, options = {}) {
    if (!this.registryLoaded) {
      await this.loadRegistry();
    }
    
    try {
      // Check if dataset exists in registry
      if (this.datasetRegistry[datasetId]) {
        console.log(`Dataset ${datasetId} found in registry`);
      }
      
      // Use base implementation to load dataset
      const result = await this.call('load_dataset', {
        dataset_id: datasetId,
        subset: options.subset,
        split: options.split
      });
      
      if (result && !result.error) {
        // Add or update dataset in registry
        this.datasetRegistry[datasetId] = {
          dataset_id: datasetId,
          subset: options.subset,
          splits: result.splits || ['train'],
          features: result.features || {},
          last_updated: Date.now()
        };
        
        // Add local path if available
        if (result.local_path) {
          this.datasetRegistry[datasetId].local_path = result.local_path;
        }
        
        // Add IPFS CID if available
        if (result.ipfs_cid) {
          this.datasetRegistry[datasetId].ipfs_cid = result.ipfs_cid;
        }
        
        // Save registry
        await this.saveRegistry();
      }
      
      return result;
    } catch (error) {
      console.error(`Failed to load dataset ${datasetId}:`, error);
      return { error: error.message };
    }
  }
  
  /**
   * Import a dataset from IPFS
   * @param {string} datasetId - Dataset identifier
   * @param {string} cid - IPFS content identifier
   */
  async importDatasetFromIPFS(datasetId, cid) {
    if (!this.registryLoaded) {
      await this.loadRegistry();
    }
    
    if (!this.ipfsKit) {
      return { error: 'IPFS Kit not available' };
    }
    
    try {
      // Call the base implementation
      const result = await this.call('import_dataset_from_ipfs', {
        dataset_id: datasetId,
        cid: cid
      });
      
      if (result && !result.error) {
        // Add to registry
        this.datasetRegistry[datasetId] = {
          dataset_id: datasetId,
          ipfs_cid: cid,
          local_path: result.local_path,
          splits: result.splits || ['train'],
          features: result.features || {},
          last_updated: Date.now(),
          source: 'ipfs'
        };
        
        // Save registry
        await this.saveRegistry();
      }
      
      return result;
    } catch (error) {
      console.error(`Failed to import dataset ${datasetId} from IPFS:`, error);
      return { error: error.message };
    }
  }
  
  /**
   * Remove a dataset
   * @param {string} datasetId - Dataset identifier
   */
  async removeDataset(datasetId) {
    if (!this.registryLoaded) {
      await this.loadRegistry();
    }
    
    if (!this.datasetRegistry[datasetId]) {
      console.warn(`Dataset ${datasetId} not found in registry`);
      return false;
    }
    
    try {
      const result = await this.call('remove_dataset', {
        dataset_id: datasetId
      });
      
      if (result && result !== false) {
        // Remove from registry
        delete this.datasetRegistry[datasetId];
        
        // Save registry
        await this.saveRegistry();
        
        return true;
      }
      
      return result;
    } catch (error) {
      console.error(`Failed to remove dataset ${datasetId}:`, error);
      return false;
    }
  }
  
  /**
   * Get sample rows from a dataset
   * @param {string} datasetId - Dataset identifier
   * @param {string} split - Dataset split
   * @param {number} numSamples - Number of samples to return
   */
  async getSample(datasetId, split = null, numSamples = 5) {
    try {
      const result = await this.call('get_sample', {
        dataset_id: datasetId,
        split: split,
        num_samples: numSamples
      });
      
      return result;
    } catch (error) {
      console.error(`Failed to get sample from dataset ${datasetId}:`, error);
      return { error: error.message };
    }
  }
  
  /**
   * Run module tests
   */
  async test() {
    console.log('Testing IPFS Datasets module');
    
    try {
      // Test initialization
      let testResults = {
        success: true,
        module: 'datasets',
        initialization: false,
        registry: false,
        list_datasets: false,
        dataset_loading: false,
        sample_loading: false,
        ipfs_integration: false,
        capabilities: {
          huggingface_datasets: false,
          ipfs: this.ipfsKit !== null,
          ipfs_datasets_js: true
        },
        metadata: this.metadata
      };
      
      // Test initialization
      const initResult = await this.init();
      testResults.initialization = initResult;
      
      // Test registry
      const registryResult = await this.loadRegistry();
      testResults.registry = registryResult;
      
      // Test listing datasets
      const datasets = await this.listDatasets();
      testResults.list_datasets = typeof datasets === 'object';
      
      // Test dataset loading and sampling via the base implementation
      try {
        const testDataset = await this.call('test_dataset_loading', {});
        testResults.dataset_loading = testDataset && !testDataset.error;
        
        if (testResults.dataset_loading) {
          const sampleResult = await this.call('test_sample_loading', {});
          testResults.sample_loading = sampleResult && !sampleResult.error;
        }
      } catch (error) {
        console.error('Dataset loading test failed:', error);
      }
      
      // Test IPFS integration
      if (this.ipfsKit) {
        try {
          const ipfsTest = await this.call('test_ipfs_integration', {});
          testResults.ipfs_integration = ipfsTest && !ipfsTest.error;
        } catch (error) {
          console.error('IPFS integration test failed:', error);
        }
      }
      
      // Update success based on individual test results
      testResults.success = testResults.initialization && 
                          testResults.registry && 
                          testResults.list_datasets;
      
      return testResults;
    } catch (error) {
      console.error('IPFS Datasets test failed:', error);
      return {
        success: false,
        module: 'datasets',
        error: error.message,
        initialization: false,
        registry: false,
        list_datasets: false,
        dataset_loading: false,
        sample_loading: false,
        ipfs_integration: false,
        capabilities: {
          huggingface_datasets: false,
          ipfs: this.ipfsKit !== null,
          ipfs_datasets_js: true
        },
        metadata: this.metadata
      };
    }
  }
}

// Create default instance
const ipfsDatasets = new IPFSDatasets();

export { IPFSDatasets, ipfsDatasets };
export default ipfsDatasets;

File: hallucinate_app/python/hallucinate_app/ipfs_datasets.py
"""
IPFS Datasets Module

This module provides a wrapper around the ipfs_datasets_py package, integrating it
with the hallucinate_app resource pool pattern and adding hallucinate_app-specific
functionality.
"""

import os
import sys
import json
import logging
import asyncio
import time
from typing import Dict, List, Any, Optional, Union, Callable
from hallucinate_app.submodule_compat import instantiate_from_candidates

# Configure logging
logging.basicConfig(level=logging.INFO, 
                    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

# Try to import ipfs_datasets_py
try:
    import ipfs_datasets_py
    from ipfs_datasets_py import load_dataset
    HAVE_DATASETS = True
except ImportError:
    logger.warning("ipfs_datasets_py package not found. Some features will be limited.")
    HAVE_DATASETS = False

# Try to import datasets (HuggingFace)
try:
    import datasets
    HAVE_HUGGINGFACE = True
except ImportError:
    logger.warning("HuggingFace datasets package not found. Using limited functionality.")
    HAVE_HUGGINGFACE = False


class IPFSDatasets:
    """
    IPFS Datasets Manager for hallucinate_app
    
    Provides dataset management with IPFS integration, leveraging the ipfs_datasets_py package.
    """
    
    def __init__(self, resources: Dict[str, Any] = None, metadata: Dict[str, Any] = None):
        """
        Initialize the IPFS Datasets Manager
        
        Args:
            resources: Resource pool for accessing other modules
            metadata: Configuration metadata
        """
        self.resources = resources or {}
        self.metadata = metadata or {}
        
        # Configuration with defaults
        self.config = {
            "dataset_cache_dir": self.metadata.get("dataset_cache_dir", "./datasets"),
            "use_caching": self.metadata.get("use_caching", True),
            "max_memory": self.metadata.get("max_memory", "1GB")
        }
        
        # State tracking
        self.initialized = False
        self.datasets_manager = None
        
        # Dataset tracking
        self.loaded_datasets = {}
        self.integration_metrics = {
            "load_dataset_calls": 0,
            "load_dataset_errors": 0,
            "load_dataset_total_ms": 0.0,
        }
        
        logger.info("IPFS Datasets module initialized with configuration")
    
    async def init(self) -> bool:
        """
        Initialize the IPFS Datasets Manager
        
        Returns:
            bool: True if initialization successful
        """
        try:
            # Check if required dependencies are installed
            if not HAVE_DATASETS:
                logger.warning("ipfs_datasets_py package not installed, functionality will be limited")
            
            # Ensure we have IPFS Kit available
            if "ipfs_kit" not in self.resources:
                logger.warning("IPFS Kit resource not available, using standalone mode")
            
            # Initialize any required directories
            os.makedirs(self.config["dataset_cache_dir"], exist_ok=True)
            
            # Initialize the underlying ipfs_datasets_py instance if available
            if HAVE_DATASETS:
                self.datasets_manager = instantiate_from_candidates(
                    ipfs_datasets_py,
                    ("ipfs_datasets_py", "IPFSDatasetsPy", "ipfs_datasets"),
                    self.resources,
                    self.metadata
                )
            
            self.initialized = True
            logger.info("IPFS Datasets module initialized successfully")
            return True
        except Exception as e:
            logger.error(f"Failed to initialize IPFS Datasets: {e}")
            return False
    
    async def load_dataset(self, dataset_name: str, split: Optional[str] = None) -> Dict[str, Any]:
        """
        Load a dataset from HuggingFace or local cache
        
        Args:
            dataset_name: Name of the dataset
            split: Dataset split to load
            
        Returns:
            Dictionary with dataset info
        """
        if not self.initialized:
            await self.init()
            
        try:
            started_at = time.perf_counter()
            self.integration_metrics["load_dataset_calls"] += 1
            # Create dataset key for tracking
            dataset_key = f"{dataset_name}:{split}" if split else dataset_name
            
            # Check if already loaded
            if dataset_key in self.loaded_datasets:
                logger.info(f"Using cached dataset: {dataset_key}")
                return {
                    "success": True, 
                    "dataset": dataset_name, 
                    "split": split,
                    "cached": True
                }
            
            if HAVE_DATASETS:
                # Use ipfs_datasets_py implementation
                await self.datasets_manager.load_dataset(dataset_name, split)
                self.loaded_datasets[dataset_key] = {
                    "name": dataset_name,
                    "split": split,
                    "loaded_at": self._get_timestamp()
                }
                return {
                    "success": True, 
                    "dataset": dataset_name, 
                    "split": split
                }
            elif HAVE_HUGGINGFACE:
                # Fallback to direct HuggingFace loading
                if split:
                    dataset = datasets.load_dataset(dataset_name, split=split)
                else:
                    dataset = datasets.load_dataset(dataset_name)
                
                self.loaded_datasets[dataset_key] = {
                    "name": dataset_name,
                    "split": split,
                    "dataset": dataset,
                    "loaded_at": self._get_timestamp()
                }
                
                return {
                    "success": True, 
                    "dataset": dataset_name, 
                    "split": split,
                    "direct_hf": True
                }
            else:
                raise ImportError("Neither ipfs_datasets_py nor datasets is available")
                
        except Exception as e:
            self.integration_metrics["load_dataset_errors"] += 1
            logger.error(f"Error loading dataset {dataset_name}: {e}")
            return {
                "success": False, 
                "error": str(e),
                "dataset": dataset_name,
                "split": split
            }
        finally:
            elapsed_ms = (time.perf_counter() - started_at) * 1000.0
            self.integration_metrics["load_dataset_total_ms"] += elapsed_ms

    async def get_integration_metrics(self) -> Dict[str, Any]:
        avg_ms = 0.0
        if self.integration_metrics["load_dataset_calls"]:
            avg_ms = self.integration_metrics["load_dataset_total_ms"] / self.integration_metrics["load_dataset_calls"]
        return {
            "success": True,
            "metrics": {
                **self.integration_metrics,
                "load_dataset_avg_ms": round(avg_ms, 2),
            }
        }
    
    async def list_datasets(self) -> Dict[str, Any]:
        """
        List all currently loaded datasets
        
        Returns:
            Dictionary of loaded datasets
        """
        if not self.initialized:
            await self.init()
            
        return {
            "success": True,
            "datasets": self.loaded_datasets
        }
    
    async def dataset_info(self, dataset_name: str, split: Optional[str] = None) -> Dict[str, Any]:
        """
        Get information about a dataset
        
        Args:
            dataset_name: Name of the dataset
            split: Dataset split
            
        Returns:
            Dictionary with dataset information
        """
        if not self.initialized:
            await self.init()
            
        dataset_key = f"{dataset_name}:{split}" if split else dataset_name
        
        if dataset_key not in self.loaded_datasets:
            return {
                "success": False,
                "error": f"Dataset {dataset_key} not loaded"
            }
        
        try:
            if HAVE_DATASETS:
                # Get info using ipfs_datasets_py implementation
                dataset_info = {
                    **self.loaded_datasets[dataset_key]
                }
                
                return {
                    "success": True,
                    "info": dataset_info
                }
            elif HAVE_HUGGINGFACE:
                # Get info directly from loaded HuggingFace dataset
                dataset = self.loaded_datasets[dataset_key]["dataset"]
                info = {
                    "name": dataset_name,
                    "split": split,
                    "features": str(dataset.features) if hasattr(dataset, "features") else None,
                    "num_rows": dataset.num_rows if hasattr(dataset, "num_rows") else None,
                    "column_names": dataset.column_names if hasattr(dataset, "column_names") else None,
                    "loaded_at": self.loaded_datasets[dataset_key]["loaded_at"],
                }
                
                return {
                    "success": True,
                    "info": info
                }
            else:
                return {
                    "success": False,
                    "error": "No dataset handling implementation available"
                }
        except Exception as e:
            return {
                "success": False,
                "error": str(e)
            }
    
    async def process_dataset(self, 
                             dataset_name: str, 
                             split: Optional[str] = None,
                             output_dir: Optional[str] = None) -> Dict[str, Any]:
        """
        Process a dataset for IPFS storage
        
        Args:
            dataset_name: Name of the dataset
            split: Dataset split
            output_dir: Output directory for processed files
            
        Returns:
            Dictionary with processing results
        """
        if not self.initialized:
            await self.init()
            
        if not HAVE_DATASETS:
            return {
                "success": False,
                "error": "ipfs_datasets_py is required for dataset processing"
            }
            
        try:
            # Load the dataset if not already loaded
            dataset_key = f"{dataset_name}:{split}" if split else dataset_name
            if dataset_key not in self.loaded_datasets:
                load_result = await self.load_dataset(dataset_name, split)
                if not load_result["success"]:
                    return load_result
            
            # Determine output directory
            if output_dir is None:
                output_dir = os.path.join(self.config["dataset_cache_dir"], dataset_name.replace("/", "_"))
                os.makedirs(output_dir, exist_ok=True)
            
            # Process the dataset
            # This would normally call specific methods from ipfs_datasets_py
            # As a placeholder, we'll just indicate that processing would happen here
            
            logger.info(f"Dataset {dataset_key} would be processed to {output_dir}")
            
            return {
                "success": True,
                "dataset": dataset_name,
                "split": split,
                "output_dir": output_dir,
                "message": "Dataset processing placeholder - actual processing would happen here"
            }
            
        except Exception as e:
            logger.error(f"Error processing dataset {dataset_name}: {e}")
            return {
                "success": False,
                "error": str(e)
            }
    
    async def query_dataset(self, 
                           dataset_name: str, 
                           query: Dict[str, Any],
                           split: Optional[str] = None) -> Dict[str, Any]:
        """
        Query a dataset with a structured query
        
        Args:
            dataset_name: Name of the dataset
            query: Dictionary with query parameters
            split: Dataset split
            
        Returns:
            Dictionary with query results
        """
        if not self.initialized:
            await self.init()
            
        try:
            # Load the dataset if not already loaded
            dataset_key = f"{dataset_name}:{split}" if split else dataset_name
            if dataset_key not in self.loaded_datasets:
                load_result = await self.load_dataset(dataset_name, split)
                if not load_result["success"]:
                    return load_result
            
            # Execute the query
            if HAVE_DATASETS:
                # This would call specific methods from ipfs_datasets_py
                # As a placeholder, just return the query parameters
                return {
                    "success": True,
                    "dataset": dataset_name,
                    "split": split,
                    "query": query,
                    "results": "Query would be executed here with ipfs_datasets_py"
                }
            elif HAVE_HUGGINGFACE:
                # Execute query directly on HuggingFace dataset
                dataset = self.loaded_datasets[dataset_key]["dataset"]
                
                # Extract filtering criteria
                filter_column = query.get("filter_column")
                filter_value = query.get("filter_value")
                
                if filter_column and filter_value:
                    filtered_dataset = dataset.filter(lambda x: x[filter_column] == filter_value)
                    sample_rows = filtered_dataset[:5] if len(filtered_dataset) >= 5 else filtered_dataset
                    
                    return {
                        "success": True,
                        "dataset": dataset_name,
                        "split": split,
                        "query": query,
                        "num_results": len(filtered_dataset),
                        "sample_results": sample_rows
                    }
                else:
                    return {
                        "success": False,
                        "error": "Missing filter_column or filter_value in query"
                    }
            else:
                return {
                    "success": False,
                    "error": "No dataset query implementation available"
                }
                
        except Exception as e:
            logger.error(f"Error querying dataset {dataset_name}: {e}")
            return {
                "success": False,
                "error": str(e)
            }
    
    async def test(self) -> Dict[str, Any]:
        """
        Run self-test
        
        Returns:
            Dictionary with test results
        """
        test_results = {
            "success": True,
            "module": "ipfs_datasets",
            "tests": []
        }
        
        try:
            # Test initialization
            if not self.initialized:
                init_result = await self.init()
                test_results["tests"].append({
                    "name": "initialization",
                    "success": init_result
                })
                
                if not init_result:
                    test_results["success"] = False
                    return test_results
            
            # Test dataset loading - only if we have the required packages
            if HAVE_HUGGINGFACE:
                try:
                    # Use a small test dataset
                    dataset_result = await self.load_dataset("hf-internal-testing/dummy_dataset", "train")
                    
                    test_results["tests"].append({
                        "name": "dataset_loading",
                        "success": dataset_result["success"],
                        "details": dataset_result if not dataset_result["success"] else None
                    })
                    
                    if not dataset_result["success"]:
                        test_results["success"] = False
                except Exception as e:
                    test_results["tests"].append({
                        "name": "dataset_loading",
                        "success": False,
                        "error": str(e)
                    })
                    test_results["success"] = False
            else:
                test_results["tests"].append({
                    "name": "dataset_loading",
                    "success": False,
                    "error": "HuggingFace datasets package not available"
                })
                test_results["success"] = False
            
            # Test listing datasets
            try:
                list_result = await self.list_datasets()
                
                test_results["tests"].append({
                    "name": "list_datasets",
                    "success": list_result["success"]
                })
                
                if not list_result["success"]:
                    test_results["success"] = False
            except Exception as e:
                test_results["tests"].append({
                    "name": "list_datasets",
                    "success": False,
                    "error": str(e)
                })
                test_results["success"] = False
                
            return test_results
        except Exception as e:
            test_results["success"] = False
            test_results["error"] = str(e)
            return test_results
    
    def _get_timestamp(self) -> str:
        """Get current timestamp string"""
        from datetime import datetime
        return datetime.now().isoformat()


# Example usage
if __name__ == "__main__":
    async def main():
        # Create IPFSDatasets instance
        datasets_manager = IPFSDatasets(
            resources={},
            metadata={"dataset_cache_dir": "./dataset_cache"}
        )
        
        # Initialize
        await datasets_manager.init()
        
        # Run test
        test_results = await datasets_manager.test()
        print(json.dumps(test_results, indent=2))
        
    # Run the example
    asyncio.run(main())
